File: CodeSamples/kahn.lua
function kahn(graph, n)
    local inverse = {}
    for u, neighbors in pairs(graph) do
        for _, v in ipairs(neighbors) do
            inverse[v] = inverse[v] or {}
            table.insert(inverse[v], u)
        end
    end

    local stack = {}
    for v = 0, n - 1 do
        if not inverse[v] then
            table.insert(stack, v)
        end
    end

    local order = {}
    while #stack > 0 do
        local u = table.remove(stack, #stack)
        table.insert(order, u)
        local neighbors = graph[u] or {}
        local neighborList = {}
        for v in pairs(neighbors) do
            table.insert(neighborList, v)
        end
        for _, v in ipairs(neighborList) do
            graph[u][v] = nil
            inverse[v][u] = nil
            if not next(inverse[v]) then
                table.insert(stack, v)
            end
        end
    end

    if #order ~= n then
        return "Cyclic graph"
    end

    return order
end

local input_file = io.open("input.txt", "r")
local output_file = io.open("output.txt", "w")

local t = tonumber(input_file:read())
for i = 1, t do
    local n = tonumber(input_file:read())
    local graph = {}

    while true do
        local line = input_file:read()
        if not line then
            break
        end

        local u, v = line:match("(%S+)%s+(%S+)")
        u, v = tonumber(u), tonumber(v)

        if not graph[u] then
            graph[u] = {}
        end

        table.insert(graph[u], v)
    end

    local result = kahn(graph, n)
    if result == "Cyclic graph" then
        output_file:write("Cyclic graph\n")
    else
        output_file:write(table.concat(result, " ") .. "\n")
    end
end

input_file:close()
output_file:close()

File: CodeSamples/kahn.rs
use std::collections::{HashMap, HashSet};
use std::fs::File;
use std::io::{BufRead, BufReader, Write};

fn kahn(graph: &mut HashMap<i32, HashSet<i32>>, n: i32) -> Result<Vec<i32>, String> {
    let mut inverse: HashMap<i32, HashSet<i32>> = HashMap::new();
    for (u, neighbors) in graph.iter() {
        for &v in neighbors {
            inverse.entry(v).or_insert(HashSet::new()).insert(*u);
        }
    }

    let mut stack: Vec<i32> = (0..n).filter(|&v| !inverse.contains_key(&v)).collect();
    let mut order: Vec<i32> = Vec::new();

    while let Some(u) = stack.pop() {
        order.push(u);
        if let Some(neighbors) = graph.get(&u) {
            let neighbor_list: Vec<i32> = neighbors.iter().cloned().collect();
            for v in neighbor_list {
                graph.get_mut(&u).unwrap().remove(&v);
                inverse.get_mut(&v).unwrap().remove(&u);
                if inverse[&v].is_empty() {
                    stack.push(v);
                }
            }
        }
    }

    if order.len() != n as usize {
        return Err("Cyclic graph".to_string());
    }

    Ok(order)
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let file = File::open("input.txt")?;
    let reader = BufReader::new(file);
    let mut input_lines = reader.lines().map(|line| line.unwrap());

    let t: i32 = input_lines.next().unwrap().parse()?;
    let mut output = String::new();

    for _ in 0..t {
        let n: i32 = input_lines.next().unwrap().parse()?;
        let mut graph: HashMap<i32, HashSet<i32>> = HashMap::new();

        loop {
            let line = input_lines.next().unwrap();
            if line.is_empty() {
                break;
            }
            let values: Vec<i32> = line
                .split_whitespace()
                .map(|s| s.parse().unwrap())
                .collect();
            let u = values[0];
            let v = values[1];
            graph.entry(u).or_insert(HashSet::new()).insert(v);
        }

        match kahn(&mut graph, n) {
            Ok(result) => {
                output.push_str(&result.iter().map(|&x| x.to_string()).collect::<Vec<String>>().join(" "));
                output.push('\n');
            }
            Err(_) => output.push_str("Cyclic graph\n"),
        }
    }

    let mut output_file = File::create("output.txt")?;
    output_file.write_all(output.as_bytes())?;

    Ok(())
}
